doc/tutorials/code-generator.md

```diff
@@ -0,0 +1,272 @@
+- _template base name_: Unlike the template name, this identifier is internal to the framework. It will not be seen by users of `kdb gen`.
+  This is the base name of all mustache files belonging to the template. In most cases this will be the same as the template name.
+- _input keyset_: This keyset contains the data with which the template will be instantiated by the code-generator.
+- _parent key_: The parent key is given in the command line and defines the input keyset (the keys below the parent key).
+- _output name_: The base name for the output files as given in the command line. Suffixes may be appended, if there are multiple output
+  files.
+- _parts_: Each template may consist of multiple parts. Each part corresponds to exactly one mustache file and therefore exactly one output
+  file.
+- _part suffix_: Parts are identified by there part suffix. This suffix is appended to the template base name to obtain the name of the
+  mustache file corresponding to the part. Similarly it is appended to the output name to obtain the output file for this part. The empty
+  string `""` is a valid part suffix, although it is only recommend for templates with a single output file.
+
+All templates consist of two pieces. The set of mustache files (on for each part) and an accompanying C++ class to supply the mustache
+rendering engine with data from the input keyset.
+
+## Creating a new template
+
+In this guide we will create a basic template, that generates a single file containing a simple list of all keys in our input keyset. An
+example output from running `kdb gen` with our to-be-developed template would be:
+
+```
+user/sw/myapp/#0/current
+user/sw/myapp/#0/current/dir0
+user/sw/myapp/#0/current/dir0/subdir0/key0
+user/sw/myapp/#0/current/dir0/subdir0/key1
+user/sw/myapp/#0/current/dir1
+```
+
+As you can see, the result matches that of redirecting the stdout of `kdb ls` into a file.
+
+### Creating the mustache template
+
+The first thing we need to create is a mustache template. You will need (at least) one template file for each output file. If you make use
+of partials (see [below](#using-partials)), you of course have to write multiple files.
+
+All template files must be located in the [`src/tools/kdb/gen/templates`](/src/tools/kdb/gen/templates) folder and must be named according
+to the scheme `<template base name>.<part suffix>.mustache`.
+
+For our simple example we create the file `src/tools/kdb/gen/templates/example.txt.mustache` with the following content:
+
+```
+{{# keys }}
+{{{ name }}}
+{{/ keys }}
+```
+
+Note: we will not go into detail on how mustache templates work, for more information see e.g.
+[here](https://mustache.github.io/mustache.5.html). All of features supported by the kainjow library should be supported by our framework
+as well.
+
+Our CMke script will collect all `.mustache` files in `src/tools/kdb/gen/templates` into a header containing a `static const char *` field
```

Comment:
```suggestion
Our CMake script will collect all `.mustache` files in `src/tools/kdb/gen/templates` into a header containing a `static const char *` field
```

Reply:
Thank you for the suggestions!

Did you also try everything in the tutorials? Did everything work?

Reply:
Not yet, but I will check them later this day.